install/rctf.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+# a script to manage rCTF installations
+
+import sys, os, io
+import argparse
+import logging
+
+
+logging.basicConfig(format = '%(message)s', level = logging.DEBUG)
+
+
+# define useful functions
+
+
+def verify_privileges(euid = 0):
+    if os.geteuid() != euid:
+        logging.fatal('This command must be run as root')
+        return False
+
+    return True
+
+
+def check_file(fn):
+    # verifies a file exists
+    return os.path.isfile(fn)
+
+
+def execute(command):
+    logging.debug('Executing `%s`...' % command)
+    
+    status_code = os.system(command)
+
+    if status_code:
+        logging.warning('Command failed to execute; exited with status code %d.' % status_code)
+
+    return status_code == 0
+
+
+# create simple model of rCTF
+
+
+REPOSITORY_BRANCH = 'master' # XXX: migrate REPOSITORY_BRANCH to an envvar/argument
+
+
+class rCTF:
+    def __init__(self, install_path = '/opt/rctf/'):
+        if not install_path.endswith('/'):
+            install_path += '/'
+
+        self.install_path = install_path
+
+
+    def up(self):
+        os.chdir(self.install_path)
+        
+        if not verify_privileges():
+            return False
+
+        if not execute('docker-compose up -d'):
+            logging.fatal('Failed to start rCTF instance')
+            return False
+
+        return True
+
+    
+    def down(self):
+        os.chdir(self.install_path)
+        
+        if not verify_privileges():
+            return False
+
+        if not execute('docker-compose down'):
+            logging.fatal('Failed to stop rCTF instance')
+            return False
+        
+        return True
+    
+    
+    def upgrade(self):
+        os.chdir(self.install_path)
+        
+        if not verify_privileges():
+            return False
+    
+        # XXX: is there a way to make this not error if it fails?
+        self.down()
+        
+        if not execute('git checkout %s' % REPOSITORY_BRANCH):
+            logging.fatal('Failed to checkout repository branch %s' % REPOSITORY_BRANCH)
+            return False
+
+        if not execute('git pull'):
+            logging.fatal('Failed to pull latest from repository')
+            return False
+
+        if not execute('docker-compose build --no-cache'):
+            logging.fatal('Failed to rebuild docker image')
+            return False
+        
+        return True
+
+
+
+# main
+
+
+if __name__ == '__main__':
+    # parse arguments
+
+
+    parser = argparse.ArgumentParser(description = 'Manage rCTF installations from the CLI')
+    
+    parser.add_argument('--install-path', type = str, default = os.environ.get('RCTF_INSTALL_PATH', os.environ.get('INSTALL_PATH', '/opt/rctf/')), help = 'The path to the rCTF installation to manage')
```

Comment:
Maybe default should be current working directory? A lot of people will probably git clone and cd into rctf.

Reply:
Since the rCTF autoinstall script installs to /opt/rctf/, I wanted to prioritize that directory. I think long-term we should default to `.` (current directory) and just `sed` to replace the default value to `/opt/rctf/` from the install script. TODO.